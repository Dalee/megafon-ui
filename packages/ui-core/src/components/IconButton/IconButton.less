@import '~styles/base.less';

@b: e('.mfui-beta-icon-button');

@{b} {
    border-radius: 50%;
    padding: 0;

    &__content {
        display: block;
    }

    .generateSize(all);

    @media @desktopMU {
        .generateSize(wide);
    }

    @media @desktop {
        .generateSize(desktop);
    }

    @media @mobileB {
        .generateSize(tablet);
    }

    @media @mobileSM {
        .generateSize(mobile);
    }

    .generateSize(@resolution) {
        &_size-@{resolution}_small &__inner {
            width: 32px;

            @{b}__content {
                width: 24px;
                height: 24px;
            }
        }

        &_size-@{resolution}_medium &__inner {
            width: 48px;

            @{b}__content {
                width: 32px;
                height: 32px;
            }
        }

        &_size-@{resolution}_large &__inner {
            width: 60px;

            @{b}__content {
                width: 40px;
                height: 40px;
            }
        }
    }
}
