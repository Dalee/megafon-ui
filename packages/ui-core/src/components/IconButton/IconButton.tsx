import * as React from 'react';
import PropTypes from 'prop-types';
import cnCreate from 'utils/cnCreate';
import './IconButton.less';
import Button from 'components/Button/Button';

export interface IIconButtonProps {
    /** Цветовая тема */
    theme?: 'green' | 'purple' | 'white' | 'black';
    /** Тип отображения */
    type?: 'primary' | 'outline';
    /** Ссылка */
    href?: string;
    /** Target - аргумент тега <a> */
    target?: '_self' | '_blank' | '_parent' | '_top';
    /** Размер на всех разрешениях экрана */
    sizeAll?: 'small' | 'medium' | 'large';
    /** Размер на разрешении экрана 1280+ */
    sizeWide?: 'small' | 'medium' | 'large';
    /** Размер на разрешении экрана 1024+ */
    sizeDesktop?: 'small' | 'medium' | 'large';
    /** Размер на разрешении экрана 768-1023 */
    sizeTablet?: 'small' | 'medium' | 'large';
    /** Размер на разрешении экрана 0-767 */
    sizeMobile?: 'small' | 'medium' | 'large';
    /** Иконка */
    icon: JSX.Element;
    /** Управление возможностью взаимодействия с компонентом */
    disabled?: boolean;
    /** Обработчик клика по кнопке */
    onClick?: (e: React.SyntheticEvent<EventTarget>) => void;
}

const cn = cnCreate('mfui-beta-icon-button');
const IconButton: React.FC<IIconButtonProps> = props => {
    const {
        theme = 'green',
        type = 'primary',
        href,
        target,
        sizeAll = 'medium',
        sizeWide,
        sizeDesktop,
        sizeTablet,
        sizeMobile,
        icon,
        disabled,
        onClick,
    } = props;

    const currentTheme = React.useMemo(() => (
        (type === 'primary') && (theme === 'black') ? 'green' : theme
    ), [theme]);

    return (
        <Button
            classes={{
                root: cn(''),
                content: cn('content'),
                inner: cn('inner'),
            }}
            className={cn({
                disabled,
                'size-all': sizeAll,
                'size-wide': sizeWide,
                'size-desktop': sizeDesktop,
                'size-tablet': sizeTablet,
                'size-mobile': sizeMobile,
            })}
            disabled={disabled}
            theme={currentTheme}
            target={target}
            href={href}
            type={type}
            sizeAll={sizeAll}
            sizeWide={sizeWide}
            sizeDesktop={sizeDesktop}
            sizeTablet={sizeTablet}
            sizeMobile={sizeMobile}
            onClick={onClick}
        >
            {icon}
        </Button>
    );
};

IconButton.propTypes = {
    theme: PropTypes.oneOf(['green', 'purple', 'white', 'black']),
    type: PropTypes.oneOf(['primary', 'outline']),
    href: PropTypes.string,
    target: PropTypes.oneOf(['_self', '_blank', '_parent', '_top']),
    sizeAll: PropTypes.oneOf(['small', 'medium', 'large']),
    sizeWide: PropTypes.oneOf(['small', 'medium', 'large']),
    sizeDesktop: PropTypes.oneOf(['small', 'medium', 'large']),
    sizeTablet: PropTypes.oneOf(['small', 'medium', 'large']),
    sizeMobile: PropTypes.oneOf(['small', 'medium', 'large']),
    icon: PropTypes.element.isRequired,
    disabled: PropTypes.bool,
    onClick: PropTypes.func,
};

export default IconButton;
