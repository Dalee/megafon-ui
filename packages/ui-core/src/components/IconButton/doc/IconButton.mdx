---
name: IconButton
menu: Компоненты
route: components/iconbutton
---

import { Playground, Props } from 'docz';
import IconButton from '../IconButton';
import Example from './IconButton.example.mdx';

# IconButton

<CopyToClipboardBox importPath="import { IconButton } from '@megafon/ui-core';" />

<DoczTabs>
    <Example />
    <Props of={IconButton} />
</DoczTabs>
