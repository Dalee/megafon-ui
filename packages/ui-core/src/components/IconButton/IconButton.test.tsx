import React from 'react';
import { shallow } from 'enzyme';
import IconButton, { IIconButtonProps } from './IconButton';
import Icon from 'icons/System/16/Gag_16.svg';

const props: IIconButtonProps = {
    className: 'customClassName',
    theme: 'purple',
    type: 'primary',
    href: 'href',
    target: '_blank',
    sizeAll: 'large',
    sizeMobile: 'small',
    sizeTablet: 'medium',
    sizeDesktop: 'large',
    sizeWide: 'large',
    disabled: true,
    icon: <Icon />,
};

describe('<IconButton />', () => {
    describe('snapshots', () => {
        it('renders with required props', () => {
            const wrapper = shallow(<IconButton icon={props.icon} />);

            expect(wrapper).toMatchSnapshot();
        });

        it('renders with props', () => {
            const wrapper = shallow(<IconButton {...props} />);

            expect(wrapper).toMatchSnapshot();
        });
    });

    it('handle onClick callback prop', () => {
        const onClick = jest.fn();
        const wrapper = shallow(<IconButton icon={props.icon} onClick={onClick} />);

        wrapper.simulate('click');

        expect(onClick).toBeCalled();
    });
});
