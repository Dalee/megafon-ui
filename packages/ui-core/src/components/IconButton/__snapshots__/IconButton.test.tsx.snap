// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<IconButton /> snapshots renders with props 1`] = `
<Button
  classes={
    Object {
      "content": "mfui-beta-icon-button__content",
      "inner": "mfui-beta-icon-button__inner",
      "root": "mfui-beta-icon-button mfui-beta-icon-button_disabled mfui-beta-icon-button_size-all_large mfui-beta-icon-button_size-wide_large mfui-beta-icon-button_size-desktop_large mfui-beta-icon-button_size-tablet_medium mfui-beta-icon-button_size-mobile_small customClassName",
    }
  }
  disabled={true}
  href="href"
  sizeAll="large"
  sizeDesktop="large"
  sizeMobile="small"
  sizeTablet="medium"
  sizeWide="large"
  target="_blank"
  theme="purple"
  type="primary"
>
  <test-file-stub />
</Button>
`;

exports[`<IconButton /> snapshots renders with required props 1`] = `
<Button
  classes={
    Object {
      "content": "mfui-beta-icon-button__content",
      "inner": "mfui-beta-icon-button__inner",
      "root": "mfui-beta-icon-button mfui-beta-icon-button_size-all_medium",
    }
  }
  sizeAll="medium"
  theme="green"
  type="primary"
>
  <test-file-stub />
</Button>
`;
